### PR TITLE
Proposal to use Souin as cache system

### PR DESCRIPTION
**Pull request by darkweak**
As we discussed on Slack, here is the PR to use [Souin](https://github.com/darkweak/souin) as cache system. It respects the [Caddy](https://github.com/caddyserver/caddy) expectations, supports JSON and `Caddyfile` configuration formats, supports Olric as embedded mode or client mode and for those who don't need a distributed storage, it supports a non distributed one.

It closes the #16 

Open to discuss with you about it.

### CODE REVIEW

File: go.mod
Comment (francislavoie) [inline go.mod:1]: This is the wrong module ID

File: main.go
Comment (francislavoie) [inline main.go:174]: I'm not seeing any way to configure olric via Caddyfile global options. That'll need to be added.

File: Caddyfile
Comment (francislavoie) [inline Caddyfile:7]: What's the purpose of this `configuration` level here? Seems useless. Shouldn't it get hoisted up a level?

Comment (Darkweak) [reply Caddyfile:7]: Because we have to separate the configuration from the `path` declaration for the file or the `url`declaration for the cluster address. IMO it's the best compromise.

Comment (francislavoie) [reply Caddyfile:7]: I don't understand - that doesn't seem like a concern that should be exposed to Caddy configuration.

Comment (Darkweak) [reply Caddyfile:7]: We could have some keys conflicts between the whole configuration (placed under `configuration`) and the file configuration (the file location is defined with the `path` one) and the cluster url/ip using the olric client (this is defined with the `url` one).

Comment (francislavoie) [reply Caddyfile:7]: I don't know what you mean because I don't see `path` or `url` in there, I only see `configuration`, which is the only element of `olric`, therefore `configuration` looks redundant.

Comment (Darkweak) [reply Caddyfile:7]: Oh, I'll add other configuration files to be more explicit :)

Comment (francislavoie) [reply Caddyfile:7]: But do we really want to let users configure it via a separate file? That seems counterproductive to me. We should keep this focused.

File: main.go
Comment (francislavoie) [inline main.go:175]: This doesn't look like a good approach. I'd call this a hack more than anything.

I think the configuration should be more explicitly done rather than relying on the existence of `}` values. No other Caddyfile parsing does that, because that should be abstracted away.

Comment (Darkweak) [reply main.go:175]: This is done to load the Olric configuration from the Caddyfile without duplicating the Olric internal type and map values over this.

File: Caddyfile
Comment (Darkweak) [reply Caddyfile:7]: Examples are available.

**Comment by Darkweak** [comment]
Update: I have to tag a new version on [Souin](https://github.com/darkweak/souin) with the embedded Olric support and the Ykeys. After that the CI will be green.

**Comment by Darkweak** [comment]
Any update about the implementation @dunglas @mholt ?

**Comment by mholt** [comment]
I don't have a strong opinion either way. I think you guys (or someone else who's invested in the cache functionality) should take charge on this feature.

**Comment by Darkweak** [comment]
@dunglas can you rerun the test suite please ?

**Comment by dunglas** [comment]
done

**Comment by Darkweak** [comment]
Can you re-run again @dunglas ? Thanks to GH to block the CI for non-maintainer contributor and require an approval before each CI run.

**Comment by Darkweak** [comment]
Should be ok now @dunglas :D

**Comment by Darkweak** [comment]
Ready to merge and tag on my side!

**Comment by mholt** [comment]
Sent you a collaborator invite -- that should give you the flexibility you need to help maintain this repo. :+1:

**Comment by Darkweak** [comment]
Any update about the implementation/review @dunglas ?

**Comment by mholt** [comment]
@Darkweak I can't speak for @dunglas, but you are now a maintainer of the repo, so you could probably merge this when you're ready.

**Comment by Darkweak** [comment]
@mholt I'm not very fan about merge without approve on a public/community repository

**Comment by dunglas** [comment]
Sorry I haven't had time to review this and I probably won't be able to before my holiday. So I agree with @mholt, let's move forward and merge.

File: Caddyfile
Comment (dunglas) [inline Caddyfile:2]: Maybe that using an ordered section (writhing brackets) would make the Caddyfile simpler?

File: README.md
Comment (dunglas) [inline README.md:9]: Is this implemented?

File: README.md
Comment (dunglas) [inline README.md:10]:
```suggestion
 * REST API to purge the cache and list stored resources
```

File: README.md
Comment (dunglas) [inline README.md:11]:
```suggestion
 * Builtin support for distributed cache
```

File: README.md
Comment (dunglas) [inline README.md:17]:
```suggestion
See the [Souin](https://github.com/darkweak/souin) configuration for the full configuration, and its associated [Caddyfile](https://github.com/darkweak/souin/plugins/caddy/Caddyfile)  
```

File: configuration.go
Comment (dunglas) [inline configuration.go:14]: This should be a `time.Duration`

File: configuration.go
Comment (dunglas) [inline configuration.go:18]: All these getters are useless and not idiomatic in Go, you can access to the field directly.

File: configuration.go
Comment (dunglas) [inline configuration.go:49]: It's not very important for now, but IMHO we shouldn't blindly copy the Varnish API and expose something modern and consistent with the rest of Caddy's APIs.

File: error.go
Comment (dunglas) [inline error.go:3]: Ad type hasn't any field, isn't a call to `errors.New()` good enough?

File: app.go
Comment (dunglas) [inline app.go:1]: Couldn't we keep the old name, which is more descriptive?

File: httpcache_test.go
Comment (dunglas) [inline httpcache_test.go:1]: IMHO we should keep the tests and check they still pass (except the one specific to the config).

File: main.go
Comment (dunglas) [inline main.go:1]: `main` is probably not a good name as it doesn't contain a `main` function.

File: main.go
Comment (dunglas) [inline main.go:71]: Recorders shouldn't be used outside of the tests.

File: main.go
Comment (dunglas) [inline main.go:134]: You can add a `return` clause here to simplify the condition.

File: main.go
Comment (dunglas) [inline main.go:234]: You should use `caddy.ParseDuration()` here.

**Review by dunglas — CHANGES_REQUESTED** [review]
I made a quick review of this part (but not of the lib itself yet).

File: configuration.go
Comment (Darkweak) [reply configuration.go:49]: We have to be compliant to be able to switch from Varnish to Souin without refactoring anything (e.g. for API P).

File: configuration.go
Comment (Darkweak) [reply configuration.go:18]: That's for the configuration interface, so it's mandatory.

File: configuration.go
Comment (Darkweak) [reply configuration.go:14]: Related to the core: https://github.com/Darkweak/Souin/issues/94

File: README.md
Comment (Darkweak) [reply README.md:9]: Yes it is.

File: main.go
Comment (Darkweak) [reply main.go:234]: Related to the core: https://github.com/Darkweak/Souin/issues/94

File: main.go
Comment (mholt) [reply main.go:71]: If it helps, here is a recorder that is more optimized for production use: https://pkg.go.dev/github.com/caddyserver/caddy/v2@v2.4.3/modules/caddyhttp#ResponseRecorder

File: configuration.go
Comment (dunglas) [reply configuration.go:49]: IMHO compatibility with the Varnish API is not a goal of this module (as compatibility with NGINX isn't a goal of Caddy). We could provide an adapter, but the native API should be in sync with current modern best practices (and the name of this header is very weird).

Comment (Darkweak) [reply configuration.go:49]: IMHO Souin has to be compliant. If it's not mandatory in Caddy we can just remove it from the documentation.

Comment (darkweak) [reply configuration.go:57]: What about this? 
Should I just exclude the Ykeys support for caddy? 
Just hide this from this caddy-module documentation and let the users free of enabling it?

**Comment by darkweak** [comment]
Wait for @dunglas feedbacks.

**Comment by polarathene** [comment]
Just curious, if this is merged to become the new `cache-handler` module, how does it differ from the caddy module from `Souin`?

Will they be effectively the same or some advantage/disadvantage between the two?

**Comment by darkweak** [comment]
@polarathene this should be equal to the caddy module from Souin.
The advantage of the Souin one would be the release delay between a new tagged version of the core and the caddy module. It will be available earlier in the Souin repository.
And there are some discussions about block some parts such as the `ykeys` support in this module which will be available in Souin repository.

**Comment by polarathene** [comment]
> this should be equal to the caddy module from Souin.

What is purpose of `cache-handler` then? Official endorsement?

Is it a simpler config/API for those that don't need more advanced full featured version of Souin caddy module?

**Comment by darkweak** [comment]
> What is the purpose of separate module cache-handler? Official endorsement under the "caddyserver" organization?

Yes that's it. 

Actually the simplest configuration for this module using Souin would be the default ttl declaration as root directive. With that, it won't enable the Souin api, won't store any ykeys to group requests by ykey.

**Comment by polarathene** [comment]
> Yes that's it.

Ah ok :)

It is to provide a simpler cache solution with smart defaults, like Caddy is well known for.

I think it would still be appreciated to better clarify the difference, than just stating it's based on Souin (since it offers it's own caddy module for more advanced features Souin has).

File: README.md
Comment (polarathene) [inline README.md:4]: Perhaps something like this? I think it makes the difference between the two clear.

I dropped the mention of "distributed" feature since that's mentioned below with other features of interest like RFC 7234 compliance.

It's not clear what the default is though, since you have both Badger (non-distributed, but can persist cache to disk?) and Olric (distributed and in-memory only), config docs can cover that or be reviewed on making it simpler (eg a `distributed` boolean instead of user needing to care about Badger or Olric providers and their configs, if they do they can change to Souin plugin).

It should be clear that distributed cache lacks on-disk persistence though (_perhaps that's not a feature `cache-handler` needs to accommodate for though?_).

```suggestion
The official Caddy HTTP cache module. This module embraces the Caddy ethos of keeping configuration simple with smart defaults out of the box!

Under the hood `cache-handler` uses the excellent HTTP caching solution [Souin](https://github.com/darkweak/souin). For advanced features like a REST API for purging cache entries or Surrogate Keys (aka YKey in Varnish), the [Souin Caddy module](https://github.com/darkweak/souin/tree/master/plugins/caddy) can be used directly when needed.
```

File: README.md
Comment (polarathene) [inline README.md:17]: Is it intended to remain compatible via config like this, or does the user need to switch to the Souin plugin to leverage the features being discussed as not supported in `cache-handler`?

File: README.md
Comment (polarathene) [inline README.md:8]: The TODO doesn't list any missing HTTP cache header support, does it no longer apply and should be removed?

```suggestion
 * Full support for HTTP cache headers defined in [RFC 7234](https://httpwg.org/specs/rfc7234.html).
```

or

```suggestion
 * [RFC 7234](https://httpwg.org/specs/rfc7234.html) compliant HTTP Cache.
```

File: README.md
Comment (polarathene) [inline README.md:14]: Once decisions on feature set are decided, a snippet Caddyfile example in the README is usually better to present to users to get a feel for it.

Links can demonstrate more verbose setups/features.

**Review by polarathene — CHANGES_REQUESTED** [review]

File: README.md
Comment (darkweak) [reply README.md:8]: I think the second one is better.

File: README.md
Comment (darkweak) [reply README.md:4]: The single `distributed` boolean option was an option but it was dropped to let the user decide to configure Olric by a configuration file or providing directly the config inside the Caddyfile to set max memory and other Olric options.

Comment (darkweak) [reply README.md:4]: I'm not sure about the `excellent` mention 😄

Comment (darkweak) [reply README.md:4]: > For advanced features like a REST API for purging cache entries or Surrogate Keys (aka YKey in Varnish), the [Souin Caddy module](https://github.com/darkweak/souin/tree/master/plugins/caddy) can be used directly when needed.
In fact the REST API is available inside the `cache-handler` but it's not enabled by default due to security reasons. And the `ykey` support may be possible (IMHO it should stay in both modules) but it won't be referenced in the `cache-handler` documentation.

File: README.md
Comment (darkweak) [reply README.md:17]: I think I should provide a minimal configuration for Caddy in the main Caddyfile (the one on at the root project directory).

**Comment by darkweak** [comment]
🆙 @dunglas @mholt @francislavoie can you re-review while I'm re-implementing the tests please ?
Thanks 🙂

**Comment by mholt** [comment]
Am pretty busy (and haven't been following this branch toooo closely, I admit) -- but let me know if there's something specific I can help out with.

**Comment by darkweak** [comment]
@dunglas tests are reimplemented

File: httpcache_test.go
Comment (dunglas) [inline httpcache_test.go:46]: This should be optional. We should use a heuristic (https://httpwg.org/specs/rfc7234.html#rfc.section.4.2.2) if a specific TTL isn't set. Varnish uses 120s by default.

File: httpcache_test.go
Comment (dunglas) [inline httpcache_test.go:117]: We should keep the previous test which is more accurate.

File: app.go
Comment (dunglas) [inline app.go:26]: You should at least inject the zap logger in Souin isn't it?

Also, I don't understand how Souin is started. It's a side effect?

File: httpcache_test.go
Comment (francislavoie) [reply httpcache_test.go:1]: We always use the same package in Caddy for tests :man_shrugging: allows to test unexported functions directly if necessary

File: app.go
Comment (dunglas) [reply app.go:52]: Ok got it. It's because allowing several cache config/instances would be counter-productive I guess?

File: httpcache_test.go
Comment (darkweak) [reply httpcache_test.go:34]: The ttl is calculated with the difference between the datetime when the response was stored and the current datetime rounded to the upper integer. Here we are making the two requests at the same time so the second response ttl is equal to 1.

File: httpcache_test.go
Comment (darkweak) [reply httpcache_test.go:46]: This should be related to the core then.

File: app.go
Comment (darkweak) [reply app.go:15]: The provider property isn't configured by the configuration so declare the json attribute isn't necessary here.

File: httpcache_test.go
Comment (darkweak) [reply httpcache_test.go:117]: On the first response we must not set the Age header. The Age header must be set when the response is served from the cache. https://datatracker.ietf.org/doc/html/rfc7234#section-4.2.3

File: app.go
Comment (mholt) [reply app.go:32]: (Not sure without a deeper look, but) it's possible that these fields aren't ready until the app is started. I know the `tls` app doesn't get to verify some things until after it's started since other apps may modify its config at Provision-time.

But yeah, normally most modules would do verification in the Validate() step.

File: app.go
Comment (mholt) [reply app.go:52]: Yeah, I think a "singleton" pattern for the cache makes sense so you're not duplicating cache values and such.

File: app.go
Comment (darkweak) [reply app.go:32]: The `Validate()` function is executed before the module provisioning and this make a nil DefaultCache property if we can this into the validate.

Comment (mholt) [reply app.go:32]: Validate() is executed _after_ Provision(), but at the same "phase" of the config load with provisioning.

Comment (darkweak) [reply app.go:32]: But if I try to check through the `Validate` method inside the App context, DefaultCache is not already populated from the runners.

Comment (francislavoie) [reply app.go:32]: I think the idea is that `Provision()` should be filling `s.DefaultCache`.

Comment (mholt) [reply app.go:32]: Oh, right, Validate() may be called before _other_ modules' Provision() is called, but is always called after the same modules' Provision().

File: httpcache_test.go
Comment (darkweak) [reply httpcache_test.go:1]: e.g. https://github.com/caddyserver/caddy/blob/master/replacer_test.go

File: app.go
Comment (darkweak) [reply app.go:26]: You may want different loglevel on each directive.

File: httpcache_test.go
Comment (dunglas) [reply httpcache_test.go:1]: Ok, as you wish (it's considered best practice to use test packages and golangci-lint can even test this automatically: https://golangci-lint.run/usage/linters/#testpackage)

File: httpcache_test.go
Comment (dunglas) [reply httpcache_test.go:34]:
```
2.4.  The ttl parameter

   "ttl" indicates the response's remaining freshness lifetime (see
   [HTTP-CACHING], Section 4.2.1) as calculated by the cache, as an
   integer number of seconds, measured as closely as possible to when
   the response header section is sent by the cache.  This includes
   freshness assigned by the cache; e.g., through heuristics (see
   [HTTP-CACHING], Section 4.2.2), local configuration, or other
   factors.  May be negative, to indicate staleness.
```

https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-cache-header#section-2.4

File: httpcache_test.go
Comment (dunglas) [reply httpcache_test.go:117]: I don't see where it's stated that the header must not be set when not coming from the cache and IIRC Varnish always set it.

Anyway the name of this test is `TestExistingAgeHeader`, so it what it must do (trigger a second request if necessary).

File: httpcache.go
Comment (dunglas) [inline httpcache.go:67]: More cases can be excluded in `plugin.CanHandle()` isn't it? Maybe could you use the `cachecontrol` library as we planed to in this module in this function as it's already a dependency of Souin? This will prevent the allocation potentially done by the buffer pool.

Comment (dunglas) [reply httpcache.go:67]: You should also set the `Cache-Status` header here to hint that it's a miss.

File: httpcache.go
Comment (dunglas) [inline httpcache.go:126]: If there is nothing to validate you can remove this method and the corresponding guard.

File: httpcache_test.go
Comment (darkweak) [reply httpcache_test.go:34]: I opened an issue on Souin repo https://github.com/darkweak/souin/issues/150

File: httpcache_test.go
Comment (darkweak) [reply httpcache_test.go:117]: I'll add the second request.

File: httpcache.go
Comment (darkweak) [reply httpcache.go:67]: You're right, I'll update the core asap.

Comment (darkweak) [reply httpcache.go:67]: So this won't prevent the allocation done by the buffer pool because we have to attach the cache-status on the response

File: Caddyfile
Comment (francislavoie) [inline Caddyfile]: Probably shouldn't remove this from here, since this file was meant as a "all options" example. You can make this `tls 120s` to match the default.

Also, I think the README should actually have the Caddyfile syntax instead of a link to a file, plus it should have a little explanation for each of the options to explain what they do and what kinds of options they take. Refer to the official Caddy docs to get an idea of how it should look, but basically a code block, then a bullet list for each option.

Comment (darkweak) [reply Caddyfile]: I will revert that and write the README this today 👍

Comment (darkweak) [reply Caddyfile]: Done ✅

**Comment by darkweak** [comment]
Ping @dunglas for the last review ?

**Comment by amanintech** [comment]
Any new update when is this going to be out ? Seems like this is the only solutions where all nice helps have headed to

**Comment by darkweak** [comment]
We are waiting for a last code review but we don't know when @dunglas will have time.

File: httpcache_test.go
Comment (dunglas) [reply httpcache_test.go:45]: As it been fixed in core?

**Review by dunglas — APPROVED** [review]
LGTM. It could be nice to handle the few remaining comments but I think it can be merged as-is to move forward.

File: httpcache_test.go
Comment (darkweak) [reply httpcache_test.go:45]: Yes it is :)

**Comment by francislavoie** [comment]
If @dunglas is good with this, then let's do it!!

This has been a long time coming. Thanks for all the effort @darkweak 😁

**Comment by mholt** [comment]
Excellent work everyone, I'm excited to see this go forward! I'll tweet about it when I have a chance.

**Comment by mattvb91** [comment]
Hi everyone thanks for getting this in!

I think im currently stuck trying to get the following config pieces in (using json api on caddy):

```
@souin-api path /souin-api*

cache @souin-api {}
```

So that I can then use that souin endpoint to purge keys. Is there anyone that could help me how to convert that into the json config? This is what i currently have:

```php
"apps"  => [
                "cache" => [
                    'api'       => [
                        'basepath' => '/cache/',
                    ],
                    "log_level" => "info",
                    'ttl'       => '3600s',
                ],
]
```
But I am not sure where the actual souin endpoint is located at now so I can purge / see the keys in cache.

I have an issue open here: #26 would appreciate any pointers! Thanks

**Pull request merged**
